subsys/net/lib/zzhc/zzhc.c

```diff
@@ -481,7 +558,8 @@ static int fsm(struct zzhc *ctx)
 		}
 
 		if (check_iccid(ctx) == 0 ||
-			check_simcard_whitelist(ctx) != 0) {
+		    check_simcard_whitelist(ctx) != 0 ||
+		    check_and_set_dns(ctx) != 0) {
```

Comment:
There will be another redundant round of DNS lookup in resolve_and_connect() if backup DNS server is used. Actually I think it'd be better to implement the logic of check_and_set_dns() in resolve_and_connect().

Reply:
@junqingzou I'm aware but for now I believe it's better to do it here early and keep it localized in one function. I'll keep thinking about it, if the positives outdo the negatives I will move it around.

Reply:
We do not have the environment to test the backup DNS anymore, which was the case in certification phase.
Please check in your code first, while I'll make further update to include missing parts from CT spec.